keps/sig-network/2732-np-version/README.md

```diff
@@ -0,0 +1,803 @@
+  implemented by plugins or not (though there is generally bidirectional
+  influence between the two).
+
+- Defining "metric-like" NetworkPolicy status information (eg, how long
+  a particular rule took to implement).
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Testing Features of Different NetworkPolicy Implementations
+
+As a Kubernetes developer (who might be named Jay), I want to create a
+NetworkPolicy test suite, and test the NetworkPolicy implementations of
+various network plugins, but I don't want to have to manually keep track
+of which plugins (and which versions of which plugins) are expected to
+correctly implement which NetworkPolicy features.
+
+#### Story 2 - More Reliable NetworkPolicy Test Cases
+
+As a developer, I want NetworkPolicy e2e test cases that work reliably,
+meaning I want the test code to be able to tell at what point a
+newly-created NetworkPolicy is expected to be in effect, so that I don't
+time out the test case before the network plugin has managed to process
+the policy.
+
+#### Story 3 - Understanding Whether New Features Are Supported
+
+As a user, I want to be able to tell whether the cluster I am working
+in implements certain new NetworkPolicy features, such as
+[NetworkPolicy port ranges] or [selecting Namespaces by name].
+
+[NetworkPolicy port ranges]: https://github.com/kubernetes/enhancements/pull/2090
+[selecting Namespaces by name]: https://github.com/kubernetes/enhancements/pull/2113
+
+#### Story 4 - Reporting NetworkPolicy Version
+
+As a network plugin developer, I want to verify that a CNI implements a specific
+NetworkPolicy version.
+
+#### Story 5 - Changing an Alpha NetworkPolicy API
+
+As a kubernetes developer, I want to make a change to a previously-added
+alpha NetworkPolicy feature, in a way that does not cause problems for
+users in clusters with varying apiserver and network plugin versions.
+
+### Notes/Constraints/Caveats
+
+#### Feature Gates and NetworkPolicy
+
+Alpha APIs create an additional problem for NetworkPolicy, due to the
+three-way (user/apiserver/plugin) versioning split; if an alpha field
+is added in 1.21 but then changed in 1.22, then a 1.22 cluster with a
+network plugin implementing the 1.21 version would probably be unable
+to use either version of the feature. We might want to establish a
+somewhat more restrictive set of alpha API rules for NetworkPolicy
+changes than are used for other API types.
+
+#### Determining Whether a Network Plugin Implements NetworkPolicyStatus
+
+If a user is waiting to see the status of a newly-created NetworkPolicy,
+there is no entirely-reliable way to distinguish "the plugin has not yet
+set `status` but will soon" from "the plugin doesn't know about `status`
+and is never going to set it".
+
+It's not clear how big a problem this is, especially if we suggest that
+implementations should create an "empty" `status` right away if it's
+going to take them a while to determine the final `status`.
```

Comment:
Kubernetes has a deprecated ComponentStatus API.

If we brought back something similar but as an official CRD, it could allow CNI plugins to report their status including what policy is supported.